user - consistently create user home directory on Linux

rpluem-vf opened the pull request:
Only create the parent of the home directory, but not the directory itself
as this will cause the following execution of luseradd in case that local
is set to yes to fail

##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
user

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

The following playbook causes the below output:

```yaml
---
- hosts: localhost
  tasks:
    - name: Create local user with non existing parent to home directory
      user:
        name: test_user
        home: /tmp/notexisting/user
        local: yes
```

<!--- Paste verbatim command output below, e.g. before and after your change -->
```paste below
ansible-playbook 2.9.13
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/root/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3.6/site-packages/ansible
  executable location = /usr/bin/ansible-playbook
  python version = 3.6.8 (default, Apr 16 2020, 01:36:27) [GCC 8.3.1 20191121 (Red Hat 8.3.1-5)]
Using /etc/ansible/ansible.cfg as config file
setting up inventory plugins
host_list declined parsing /etc/ansible/hosts as it did not pass its verify_file() method
script declined parsing /etc/ansible/hosts as it did not pass its verify_file() method
auto declined parsing /etc/ansible/hosts as it did not pass its verify_file() method
Parsed /etc/ansible/hosts inventory source with ini plugin
Loading callback plugin default of type stdout, v2.0 from /usr/lib/python3.6/site-packages/ansible/plugins/callback/default.py

PLAYBOOK: sampl.yml ***********************************************************************************************************************************************************************************************
Positional arguments: sampl.yml
verbosity: 4
connection: smart
timeout: 10
become_method: sudo
tags: ('all',)
inventory: ('/etc/ansible/hosts',)
forks: 5
1 plays in sampl.yml

PLAY [localhost] **************************************************************************************************************************************************************************************************

TASK [Gathering Facts] ********************************************************************************************************************************************************************************************
task path: /home/ruediger/zw/sampl.yml:2
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: ruediger
<127.0.0.1> EXEC /bin/sh -c 'echo ~root && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /root/.ansible/tmp `"&& mkdir "` echo /root/.ansible/tmp/ansible-tmp-1601059850.2602003-90076-99564397968919 `" && echo ansible-tmp-1601059850.2602003-90076-99564397968919="` echo /root/.ansible/tmp/ansible-tmp-1601059850.2602003-90076-99564397968919 `" ) && sleep 0'
Using module file /usr/lib/python3.6/site-packages/ansible/modules/system/setup.py
<127.0.0.1> PUT /root/.ansible/tmp/ansible-local-90069jg4om_h2/tmpe_8pop1e TO /root/.ansible/tmp/ansible-tmp-1601059850.2602003-90076-99564397968919/AnsiballZ_setup.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /root/.ansible/tmp/ansible-tmp-1601059850.2602003-90076-99564397968919/ /root/.ansible/tmp/ansible-tmp-1601059850.2602003-90076-99564397968919/AnsiballZ_setup.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/usr/bin/python3.6 /root/.ansible/tmp/ansible-tmp-1601059850.2602003-90076-99564397968919/AnsiballZ_setup.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'rm -f -r /root/.ansible/tmp/ansible-tmp-1601059850.2602003-90076-99564397968919/ > /dev/null 2>&1 && sleep 0'
ok: [localhost]
META: ran handlers

TASK [Create local user with non existing parent to home directory] ***********************************************************************************************************************************************
task path: /home/ruediger/zw/sampl.yml:4
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: ruediger
<127.0.0.1> EXEC /bin/sh -c 'echo ~root && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /root/.ansible/tmp `"&& mkdir "` echo /root/.ansible/tmp/ansible-tmp-1601059853.1873066-90160-121417519298591 `" && echo ansible-tmp-1601059853.1873066-90160-121417519298591="` echo /root/.ansible/tmp/ansible-tmp-1601059853.1873066-90160-121417519298591 `" ) && sleep 0'
Using module file /usr/lib/python3.6/site-packages/ansible/modules/system/user.py
<127.0.0.1> PUT /root/.ansible/tmp/ansible-local-90069jg4om_h2/tmpfpq4ky8b TO /root/.ansible/tmp/ansible-tmp-1601059853.1873066-90160-121417519298591/AnsiballZ_user.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /root/.ansible/tmp/ansible-tmp-1601059853.1873066-90160-121417519298591/ /root/.ansible/tmp/ansible-tmp-1601059853.1873066-90160-121417519298591/AnsiballZ_user.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/usr/bin/python3.6 /root/.ansible/tmp/ansible-tmp-1601059853.1873066-90160-121417519298591/AnsiballZ_user.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'rm -f -r /root/.ansible/tmp/ansible-tmp-1601059853.1873066-90160-121417519298591/ > /dev/null 2>&1 && sleep 0'
fatal: [localhost]: FAILED! => {
    "changed": false,
    "invocation": {
        "module_args": {
            "append": false,
            "authorization": null,
            "comment": null,
            "create_home": true,
            "expires": null,
            "force": false,
            "generate_ssh_key": null,
            "group": null,
            "groups": null,
            "hidden": null,
            "home": "/tmp/notexisting/user",
            "local": true,
            "login_class": null,
            "move_home": false,
            "name": "test_user",
            "non_unique": false,
            "password": null,
            "password_lock": null,
            "profile": null,
            "remove": false,
            "role": null,
            "seuser": null,
            "shell": null,
            "skeleton": null,
            "ssh_key_bits": 0,
            "ssh_key_comment": "ansible-generated on gauss",
            "ssh_key_file": null,
            "ssh_key_passphrase": null,
            "ssh_key_type": "rsa",
            "state": "present",
            "system": false,
            "uid": null,
            "update_password": "always"
        }
    },
    "msg": "Error creating /tmp/notexisting/user: Error creating `/tmp/notexisting/user': File exists.\n",
    "name": "test_user",
    "rc": 7
}

PLAY RECAP ********************************************************************************************************************************************************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0   

```

rpluem-vf committed:
Only create the parent of the home directory

Only create the parent of the home directory, but not the directory itself
as this will cause the following execution of luseradd in case that local
is set to yes to fail

rpluem-vf commented:
+label affects_2.9

rpluem-vf committed:
luseradd and useradd behave differently

luseradd and useradd behave differently with respect to the default
group on OpenSUSE.

rpluem-vf commented:
ready_for_review

rpluem-vf committed:
Correctly format the luseradd command

rpluem-vf committed:
Move tests in a separate file

rpluem-vf commented:
ready_for_review

rpluem-vf commented:
ready_for_review

rpluem-vf commented:
Is there anything I can do to move this forward?

rpluem-vf commented:
Any update?

samdoran commented on lib/ansible/modules/user.py:
```diff
@@ -642,7 +642,10 @@ def create_user_useradd(self):
                 if not os.path.isdir(parent):
-                    self.create_homedir(self.home)
+                    try:
+                        os.makedirs(parent)
```
This fix should be in `create_homedir()`, or add the `-M` option to the command if `self.local`.

rpluem-vf replied:
Given your comment I made some further thoughts on the general approach for creating the home directory. If we leave alone the problem that caused this PR to be created (a missing parent directory that causes `luseradd` to fail currently) I see the following situation if `create_home` is set to `yes`:
1. If the user does not exist the home directory is created and filled with the configured skeleton contents by `useradd` / `luseradd`.
2. If the user exists and just the home directory is missing then `create_homedir()` performs this job to create the home directory and copy the skeleton contents. I guess this is the cause because `usermod` / `lusermod` do not offer this functionality.

I am not 100% sure if the home directory creation and skeleton copy by `useradd` / `luseradd` always deliver the same result as `create_homedir()`. I would not even be sure if `useradd` and `luseradd` always do the same.
Hence wouldn't it be better to
1. Never use the `-k` parameter on `useradd` / `luseradd`.
2. Never use the `-m` parameter for `useradd`.
3. Always set the `-M` parameter for `useradd` / `luseradd`.
4. Let `create_homedir()` always do the job of creating the home directory and copying the skeleton content if  `create_home` is set to `yes`.
This approach would also solve the issue the current patch tries to solve.

rpluem-vf replied:
Any comments?

rpluem-vf replied:
Is there anything I can do to move this forward or that is missing?

samdoran replied:
The issue of home directory creation, particularly creating arbitrary parent directories, is quite tricky. I have been thinking about this off and on for several months and am still not clear on the “best” solution here.

In general, we really try to avoid creating arbitrary parent directories due to potential security issues. Any changes/additions to the creation of parent directories is alway something that requires a lot of careful consideration.

Compounding the situation are the increasingly apparent dirreneces between `luseradd` and `useradd`. There is a growing consensus on the Core Team that this module should be split into `local_user` and `user` modules to better handle these differences. The `local` toggle continues to create myriad issues and makes the code _much_ more complex.

samdoran replied:
Sorry that's not terribly helpful. I'm working through the failure and trying to understand it a bit more in order to suggest an appropriate fix.

rpluem-vf replied:
> In general, we really try to avoid creating arbitrary parent directories due to potential security issues. Any changes/additions to the creation of parent directories is alway something that requires a lot of careful consideration.
> 
I would like to support here. Can you give me any hint which kind of potential security issues we could hit here provided their principles are already known to the general public and you can talk about them in the public? Is it related to race conditions between the parent directory creation and creating the home directory itself?

samdoran replied:
> Can you give me any hint which kind of potential security issues we could hit here provided their principles are already known to the general public and you can talk about them in the public?

It's nothing secret. It's a fairly common attack scenario, and not in any way unique to Ansible.

If a malicious user knows the path their victim intends to create, they can create one or more of those parent directories with permissions that allow them to control sub directories, or sym linked into a directory they control.

Since the victim code is blindly creating all the directories in the path (essentially `mkdir -p`) without first checking each directory in that path for suspicious permissions or ACLs, it's possible a user's home directory could end up being accessible by an attacker.

A couple ways to thwart this attack include only creating parent directories from "trusted" roots, such as `/home` or checking, for each parent in the creation tree, if they exist and what their permissions are. The later case becomes terribly complex since some systems could have valid reasons why `/home` is 777, or there may be ACLs in place on `/home` providing some group access to all sub directories. There are just too many scenarios to come up with a reliable heuristic to detect "this directory has suspicious permissions" and then stoping execution. There would be too many false positives.

rpluem-vf replied:
But if I don't have a "trusted" root the same kind of issue can happen if I just have to create the home directory itself as the parent (or a component in the way up to '/') can be under the control of the attacker, correct? I guess if just a precreated home directory is under the attackers control this can be fixed by ensuring that its ownership and permissions are correct and that the ACL are correct / cleared. But as far as I can see other modules like ansible.builtin.file do not offer to set ACL's.

But I understand the issue that any existing path segments between a "trusted" prefix and a home directory are a potential issue and that the only safe way seems to be to create any of these segments on your own and not allow any prexisting "non trusted" segments. But this creates issues with idempotency as the "trusted" prefix which would need to be supplied by the user would need to grow after our initial execution. And as you said "detecting" a trusted path either results in false positives for insecure segments or leaves holes open.

And things become even more difficult when symlinks are added to this equation.

Another idea would be that the if the user specifies a "trusted" prefix every segment in between this and the target directory either doesn't exist or needs to have specific ownerships, permissions and ACL for the creation to succeed. But I guess this would be so awful complex to configure correctly from user perspective that it would not help either.
So maybe the best way would be to not support the creation of parent directories at all and require that the user precreates them via ansible.builtin.file before in a manner that is safe and appropriate for the specific use case.

OTOH it looks to me that ansible.builtin.file fights with similar issues and does not seem to have a silver bullet either.

samdoran replied:
> So maybe the best way would be to not support the creation of parent directories at all and require that the user precreates them via ansible.builtin.file before in a manner that is safe and appropriate for the specific use case.

> OTOH it looks to me that ansible.builtin.file fights with similar issues and does not seem to have a silver bullet either.

Exactly. 😄 It's complicated and there's no easy solution.

rpluem-vf replied:
Which brings us back to the original issue: Does it make the security situation worse if we only do `os.makedirs` up to the level above the home directory and let `useradd` / `luseradd` create the home directory? Currently we create the whole path including the home directory itself either directly via  `os.makedirs` in case there is no skeleton directory or indirectly via `shutil.copytree` which uses `os.makedirs` internally.

rpluem-vf committed:
Rename test file to follow convention and include test file

samdoran reviewed (changes requested):
After looking at this more thoroughly, it seems like omitting `-k` and `-m`, and always setting`-M` in the `User` class is the best fix here (probably best to leave it alone in other classes). The behavior of `create_home()` just interferes with `useradd`/`luseradd` options.

There are some other improvements that could be made as well, but this patch provides a nice focused fix:

```patch
diff --git lib/ansible/modules/user.py lib/ansible/modules/user.py
index e30f66f33e..7c46e333e6 100644
--- lib/ansible/modules/user.py
+++ lib/ansible/modules/user.py
@@ -644,10 +644,7 @@ class User(object):
             if self.create_home:
                 parent = os.path.dirname(self.home)
                 if not os.path.isdir(parent):
-                    try:
-                        os.makedirs(parent)
-                    except OSError as e:
-                        self.module.exit_json(failed=True, msg="%s" % to_native(e))
+                    self.create_homedir(self.home)
             cmd.append('-d')
             cmd.append(self.home)
 
@@ -669,15 +666,7 @@ class User(object):
             else:
                 cmd.append(self.password)
 
-        if self.create_home:
-            if not self.local:
-                cmd.append('-m')
-
-            if self.skeleton is not None:
-                cmd.append('-k')
-                cmd.append(self.skeleton)
-        else:
-            cmd.append('-M')
+        cmd.append('-M')
 
         if self.system:
             cmd.append('-r')

```

rpluem-vf committed:
Always use create_homedir when creating a home directory

Always use create_homedir when we are asked to create a home directory
in the User class. Don't use the -m and -k parameters from
useradd / luseradd as they behave differently with respect to
preexisting home directories. Instead always specify -M to ensure
that useradd / luseradd do not try to create the home directory.

This does not change potential different behaviours in child classes
of the User class.

rpluem-vf committed:
Consider the new umask option in create_homedir as well

Consider the new umask option from #73821 in create_homedir as well as
we do not let luseradd / useradd create the home directory any longer.

rpluem-vf commented:
I changed the code a little bit differently as we need to use `create_homedir` now always due to us always specifying `-M` not just if `not os.path.isdir(parent)` is true.
Furthermore I found that #73821 did not implement the new `umask` option correctly in `create_homedir`. I guess this wasn't caught because the tests for  #73821 do not cover the case that a parent directory is missing.

rpluem-vf committed:
Fix comment indent

samdoran commented on changelogs/fragments/fix_create_home.yml:
```diff
@@ -0,0 +1,7 @@
+bugfixes:
+  - >
+     user - The user module created the home directory itself when the parent
+     directory of the home directory was missing. This causes the following
+     execution of ``luseradd`` in case that local is set to yes to fail.
+     Now always create the home directory on our own and don't ask ``luseradd``
+     or ``useradd`` to do this for us.
```
```suggestion
  - >-
     user - properly create home path on Linux when ``local: yes`` and parent directories 
     specified in ``home`` do not exist (https://github.com/ansible/ansible/pull/71952/)
```

samdoran replied:
These fragments [end up in a list](https://github.com/ansible/ansible/blob/stable-2.11/changelogs/CHANGELOG-v2.11.rst#bugfixes), so making them succinct helps readability.

samdoran commented on lib/ansible/modules/user.py:
```diff
@@ -1250,18 +1233,26 @@ def create_homedir(self, path):
+                        for line in f:
+                            m = re.match(r'^UMASK\s+(\d+)$', line)
+                            if m:
+                                umask_string = m.group(1)
```
It should be ok to stop iterating through the file once we get a match.
```suggestion
                                umask_string = m.group(1)
                                break
```

rpluem-vf replied:
Sorry for being picky here, but this is not just an optimization, but also a change in behaviour. The old code takes the last value of `UMASK` in `/etc/login.defs` and the new one the first. I made a quick check on a Centos 8 system and at least `useradd` takes the **last** value of UMASK if there are multiple entries in `/etc/login.defs`. Hence I would propose to leave the code as is but add a comment that on purpose we don't do a break after a match as we want to get the last value just like `useradd`.

samdoran replied:
That's a very good point. I haven't ever encountered an `/etc/login.defs` file with more than one `UMASK` definition so it's not something I thought of. We can just leave it as is.

samdoran reviewed (changes requested):
Change looks very good overall. Just a few minor things.

rpluem-vf committed:
Improve wording based on review

Co-authored-by: Sam Doran <sdoran@redhat.com>

samdoran commented:
@rpluem-vf Thank you for working through this with me. I really appreciated your thoughtful (and correct!) assessments.

rpluem-vf commented:
Thank you very much for the pointer.

relrod commented:
@rpluem-vf @samdoran This is a breaking change.

Home directories are only chowned to the user [if parent directories were created](https://github.com/ansible/ansible/blob/2f87b8760de7ae1db8b40769253d8055136929fb/lib/ansible/modules/user.py#L3126-L3131).

Before this change, if I did `user: name=foo home=/foo`, then `/foo` would get created and owned by `foo` (since `useradd` would chown it appropriately). After this change, it is owned by `root`.

rpluem-vf commented:
@relrod : Thanks for this observation. This is indeed an unwanted change and I will have a look.

rpluem-vf commented:
@relrod , @samdoran : Created #74731 to fix this.
